### ISSUE TITLE
ID Token support clarification

### ISSUE THREAD

**Issue by simon-canning-octopus** [ietf-wg-wimse/draft-ietf-wimse-workload-identity-practices#11]
Hi 👋, first time contributing so please direct me elsewhere if appropriate.

I believe GitHub Actions currently supports the model defined by this document and refers to it as '[Security hardening with OpenID Connect](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect)'. The tokens GitHub Actions issues are referred to as '[OIDC tokens](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)' / '[ID Tokens](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)'.

https://github.com/ietf-wg-wimse/draft-ietf-wimse-workload-identity-bcp/blob/3916596e1656d1d21108ca660bbf060a3b57b234/draft-ietf-wimse-workload-identity-bcp.md?plain=1#L204

Is there a reason/explanation behind why ID tokens must not be accepted? Is it to ensure the token (i.e., the `sub`) represents the workload identity, and not a user identity?

Aside/context: Within my organization, and our customer base, people refer to the pattern described as 'OIDC support' or 'OIDC federated identity'. We build a product in the CI/CD tooling space.

**Comment by arndt-s** [comment]
@simon-canning-octopus please excuse the late response to this issue.

The thought to restrict ID tokens was due to the common saying that OIDC ID token are meant for authentication proof and are not meant for resource servers. They are also defined in OpenID with the strong human identity focus. I'm not sure whether the restriction remains valid here or if we should loosen it. I believe there's a lot of history here and the fact that tokens cannot be differentiated (often `typ=JWT`) doesn't help. So the statement `ID tokens MUST NOT be accepted` is not actionable.

I see the following options: (multiple possible)
* Deliberately say that the present tokens are "assertions according to RFC7523 and nothing else".
* Add working that explains the situation and also describes how OIDC ID token (don't) fit into the picture.
* Keep as is and don't allow OIDC ID Tokens.
* Keep it to OAuth only and don't mention OIDC.

What do you think?

**Comment by simon-canning-octopus** [comment]
@arndt-s I agree with your assessment that current practices are assertions according to RFC7523 and nothing else. The ID token in the GitHub Actions implementation I refer is an ID token by name only. It doesn't conform to the OIDC spec as its missing a large number of claims. It also is being used outside of the intended use, as it doesn't represent a human identity authentication proof, as you mention. Therefore I agree that ID tokens must not be accepted in the true sense of what one is. 

My remaining question is what is the goal of this document? Is it to refer to existing practices, or to do define a best practice moving forward? Given the pattern I describe is becoming the defacto, at least in the CI/CD space, what's the best way to represent this pattern, even if the token naming is confusing?

**Comment by arndt-s** [comment]
The goal of this document is just to document the patterns out there and acknowledge they exist. We are adding some additional aspects like how the credential can be delivered to the workload and more to it. 

There's been a lot of discussions whether this document should be called "Best Practices" - and if it should define something new. For that we have decided that the value of describing the current practices is already there and based on the feedback and interested we collect we can work on a document that improves security and usability afterwards.

BTW, there's changes happening on RFC7523. I recommend to track that too as it certainly affects this flow here. https://datatracker.ietf.org/doc/draft-ietf-oauth-rfc7523bis/

PR #19 aims to address some of the aspects discussed here and also adds a section describing the situation with OIDC. It would be great to get your review on it!

**Comment by simon-canning-octopus** [comment]
I added a nit suggestion to #19, but overall it clarifies the concern I've raised here.
